src/include/libpmemobj++/transaction.hpp

```diff
@@ -223,16 +223,41 @@ class transaction {
 		 * End pmemobj transaction. Depending on the context
 		 * of object destruction, the transaction will
 		 * automatically be either committed or aborted.
+		 *
+		 * @throw nvml::transaction_error if the transaction got aborted
+		 * without an active exception.
 		 */
-		~automatic() noexcept
+		~automatic() noexcept(false)
 		{
 			/* manual abort or commit end transaction */
-			if (pmemobj_tx_stage() != TX_STAGE_WORK) {
-				(void)pmemobj_tx_end();
-				return;
+			switch (pmemobj_tx_stage()) {
+				case TX_STAGE_ONCOMMIT:
+					(void)pmemobj_tx_end();
+					return;
+				case TX_STAGE_ONABORT:
+					(void)pmemobj_tx_end();
+					/* transaction aborted, throw to notify
+					 */
+					if (!exceptions
+						     .new_uncaught_exception())
```

Comment:
Well, this looks a bit silly now,
                                            doesn't it?

Reply:
Yes, I know, sorry. I initially didn't want to do it as a switch, because of the excessive indentation in our code style. But I feel like this is actually easier to understand. A lot of ifs would be necessary to get this to work. I might be wrong though. Just to get things straight, you are complaining about the `operator.` being on a different line or the switch altogether?

Reply:
Not much complaining, just noticing (I hereby vote down the 80 width limit for C++ template headers ).
There is no fall-through apparently, so three `if`s are equivalent to this `switch` (and one fewer tab of indentation).
And you didn't include TX_STAGE_WORK among the cases.

Anyways, `pmemobj_tx_end` must be called either way, isn't that the case? I mean, it is called in four places in this function.

Can't we just add a dummy member variable to this class, which makes sure it is called?
```
class dummy {
 public: ~dummy() {
   pmemobj_tx_end();
 }
};
class automatic {
  private: dummy tx_caller_member_yeaaah;
  ~automatic()
   {
       ... bla bla bla ...
       throw("abandon ship!"); <- dummy is destructed anyways
   }
};
```
I'm rusty on this part in C++, I'm not sure at which point dummy's destructor would be called here, but I have guesses.

Anyways, probably this function could look simpler without all the pmemobj_tx_end calls.

Reply:
Actually, `manual`'s destructor already does that for us.
In fact, `manual`'s constructor looks rather similar to `automatic`'s constructor. Very, very similar.

Well, maybe this can work, I ain't no C++ expert, you tell me:

```
class automatic {
   private:

    manual handled_transaction;

   public:

    template <typename... L>
    automatic(obj::pool_base &pop, L &... locks) : handled_transaction(pop, locks...)
    {
       /* Nothing to do, yupee! */
    }

  ~automatic() noexcept(false)
   {
      if (exceptions.new_uncaught_exception()) /* are we here because of an exception? */
           return; /* keep unwinding the stack */
 
      /* no exception, commiting or aborting */
      if (pmemobj_tx_stage() == TX_STAGE_WORK) {
          transaction::commit();
      } else if (pmemobj_tx_stage() == TX_STAGE_ONABORT ||
             (pmemobj_tx_stage() == TX_STAGE_FINALLY && pmemobj_tx_errno() != 0))
                 throw transaction_error("transaction aborted");
      }    
   }
};
```

Reply:
It was like 4 am when I decided to write this patch - might not be... what's the word - ideal ;) That's why I said _don't merge_. I just wanted to put it out there.

As for your ideas in particular, I like them. Especially the one with using `manual`. Seems legit.